File: 03-march/16-get-biggest-three-rhombus-sums-in-a-grid/solution.java
class Solution {
    public int[] getBiggestThree(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        java.util.HashSet<Integer> seen = new java.util.HashSet<>();

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                seen.add(grid[r][c]);
                int maxSize = Math.min(Math.min(cols - 1 - c, c), (rows - 1 - r) / 2);

                for (int size = 1; size <= maxSize; size++) {
                    int total = 0;
                    int x = r;
                    int y = c;

                    for (int step = 0; step < size; step++) {
                        total += grid[x][y];
                        x++;
                        y++;
                    }
                    for (int step = 0; step < size; step++) {
                        total += grid[x][y];
                        x++;
                        y--;
                    }
                    for (int step = 0; step < size; step++) {
                        total += grid[x][y];
                        x--;
                        y--;
                    }
                    for (int step = 0; step < size; step++) {
                        total += grid[x][y];
                        x--;
                        y++;
                    }

                    seen.add(total);
                }
            }
        }

        java.util.ArrayList<Integer> sums = new java.util.ArrayList<>(seen);
        sums.sort((a, b) -> b - a);
        int limit = Math.min(3, sums.size());
        int[] ans = new int[limit];
        for (int i = 0; i < limit; i++) {
            ans[i] = sums.get(i);
        }
        return ans;
    }
}

File: 03-march/16-get-biggest-three-rhombus-sums-in-a-grid/solution.py
class Solution:
    def getBiggestThree(self, grid: list[list[int]]) -> list[int]:
        rows = len(grid)
        cols = len(grid[0])
        seen = set()

        for r in range(rows):
            for c in range(cols):
                seen.add(grid[r][c])
                max_size = min((cols - 1 - c), c, (rows - 1 - r) // 2)
                for size in range(1, max_size + 1):
                    total = 0

                    x, y = r, c
                    for _ in range(size):
                        total += grid[x][y]
                        x += 1
                        y += 1

                    for _ in range(size):
                        total += grid[x][y]
                        x += 1
                        y -= 1

                    for _ in range(size):
                        total += grid[x][y]
                        x -= 1
                        y -= 1

                    for _ in range(size):
                        total += grid[x][y]
                        x -= 1
                        y += 1

                    seen.add(total)

        return sorted(seen, reverse=True)[:3]
